src/mds/CInode.cc

```diff
@@ -3886,47 +3882,54 @@ void CInode::validate_disk_state(CInode::validated_data *results,
 
     bool _dirfrags(int rval) {
       // basic reporting setup
-      results->raw_rstats.checked = true;
-      results->raw_rstats.ondisk_read_retval = rval;
-      results->raw_rstats.passed = false; // we'll set it true if we make it
+      results->raw_stats.checked = true;
+      results->raw_stats.ondisk_read_retval = rval;
+
+      results->raw_stats.memory_value.dirstat = in->inode.dirstat;
+      results->raw_stats.memory_value.rstat = in->inode.rstat;
+      frag_info_t& dir_info = results->raw_stats.ondisk_value.dirstat;
+      nest_info_t& nest_info = results->raw_stats.ondisk_value.rstat;
+
       if (rval != 0) {
-        results->raw_rstats.error_str << "Failed to read dirfrags off disk";
-        return true;
+        results->raw_stats.error_str << "Failed to read dirfrags off disk";
+	goto next;
       }
 
       // check each dirfrag...
-      nest_info_t& sub_info = results->raw_rstats.ondisk_value;
       for (compact_map<frag_t,CDir*>::iterator p = in->dirfrags.begin();
 	   p != in->dirfrags.end();
 	   ++p) {
-        if (!p->second->is_complete()) {
-          results->raw_rstats.error_str << "dirfrag is INCOMPLETE despite fetching; probably too large compared to MDS cache size?\n";
-          return true;
-        }
-        // FIXME!!! Don't assert out on damage!
-        assert(p->second->scrub_local());
-        sub_info.add(p->second->fnode.accounted_rstat);
+	CDir *dir = p->second;
+	assert(dir->get_version() > 0);
+	nest_info.add(dir->fnode.accounted_rstat);
+	dir_info.add(dir->fnode.accounted_fragstat);
```

Comment:
...so we'll report issues above if the dirfrag is gone and fetching returned ENOENT or similar, but not if the frag is somehow damaged (eg, inconsistent rstats). That's not really good enough, we need to actually validate and report the results.